tests/unit/s2n_server_hello_test.c

```diff
@@ -173,6 +173,155 @@ int main(int argc, char **argv)
+
+        /* Verify that the downgrade is detected */
+        struct s2n_stuffer *client_stuffer = &client_conn->handshake.io;
+        EXPECT_FAILURE_WITH_ERRNO(s2n_server_hello_recv(client_conn), S2N_ERR_PROTOCOL_DOWNGRADE_DETECTED);
+
+        EXPECT_EQUAL(s2n_stuffer_data_available(client_stuffer), 0);
+
+        EXPECT_SUCCESS(s2n_config_free(client_config));
+        EXPECT_SUCCESS(s2n_config_free(server_config));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_disable_tls13());
+    }
+
+    /* Verify a TLS1.2 client can negotiate with a TLS1.3 server */
+    {
+        struct s2n_config *client_config;
+        struct s2n_connection *client_conn;
+        struct s2n_config *server_config;
+        struct s2n_connection *server_conn;
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+
+        /* The client will request TLS1.2 */
+        client_conn->client_protocol_version = S2N_TLS12;
+
+        struct s2n_stuffer *server_stuffer = &server_conn->handshake.io;
+
+        const uint32_t total = S2N_TLS_PROTOCOL_VERSION_LEN
+            + S2N_TLS_RANDOM_DATA_LEN
+            + SESSION_ID_SIZE
+            + server_conn->session_id_len
+            + S2N_TLS_CIPHER_SUITE_LEN
+            + COMPRESSION_METHOD_SIZE;
+
+        /* The server will respond with TLS1.2 even though it support TLS1.3. This is expected because */
+        /* the client only support TLS1.2 */
+        EXPECT_SUCCESS(s2n_enable_tls13());
+        server_conn->actual_protocol_version = S2N_TLS12;
+        server_conn->secure.cipher_suite = &s2n_ecdhe_ecdsa_with_aes_128_gcm_sha256;
+        EXPECT_SUCCESS(s2n_server_hello_send(server_conn));
+        EXPECT_SUCCESS(s2n_disable_tls13());
+        EXPECT_EQUAL(s2n_stuffer_data_available(server_stuffer), total);
+
+        /* Copy server stuffer to client stuffer */
+        EXPECT_SUCCESS(s2n_stuffer_copy(&server_conn->handshake.io, &client_conn->handshake.io, total));
+
+        /* Verify that a TLS12 client does not error due to the downgrade */
+        struct s2n_stuffer *client_stuffer = &client_conn->handshake.io;
+        EXPECT_SUCCESS(s2n_server_hello_recv(client_conn));
+
+        EXPECT_EQUAL(s2n_stuffer_data_available(client_stuffer), 0);
+
+        EXPECT_SUCCESS(s2n_config_free(client_config));
+        EXPECT_SUCCESS(s2n_config_free(server_config));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
```

Comment:
just for safety purposes, could we add a test if we are running a tls12 server, a tls13 server does not error out?

Reply:
Are you referring to a TLS1.2 <=> TLS1.2 test? That is the first test in this file.

Reply:
tls 1.2 server <-> tls 1.3 client

Reply:
Added in next push.